### ISSUE TITLE
[iOS][Impeller] Remove nextDrawable latency by deferring drawable acquisition.

### ISSUE THREAD

**Issue by jonahwilliams** [flutter/flutter#138490]
# Background

See also: https://github.com/flutter/flutter/issues/134959

iOS only allows applications to request two or three swapchain images ("drawables"). When running at the device framerate Flutter applications frequently block for a handful of milliseconds on the raster thread as it waits for the next available drawable. iOS documentation [recommends requesting the drawable "as late as possible"](https://developer.apple.com/library/archive/documentation/3DDrawing/Conceptual/MTLBestPracticesGuide/Drawables.html) to avoid having this wait stall rendering workloads:

> Always acquire a drawable as late as possible; preferably, immediately before encoding an on-screen render pass. A frame’s CPU work may include dynamic data updates and off-screen render passes that you can perform before acquiring a drawable.

While the UI thread workload is performed before drawable acquisition, the engine workload is not. This can reduce the available rendering time from 8ms down to less than 4ms on high frame (120hz) rate iOS devices.

![image](https://github.com/flutter/flutter/assets/8975114/8d53c3bd-25f4-41f3-be61-a919217ff28f)

Pictured Above: On a 120hz iPhone we have 8.33 ms to complete rendering (CPU and GPU workloads) of a frame and submit the drawable. According to metal system traces from  [#134959](https://github.com/flutter/flutter/issues/134959), this can take 20-25ms. Assuming constantly produced frames and ordering of drawables, this implies that frame 4 will need to wait ~4ms for drawable acquisition before it can do any work. Oftentimes this doesn't result in dropped frames, as we can usually finish the frame workload in under a ms for simple apps. But that isn't always true, and the persistently high raster times also worry our user base.

![image](https://github.com/flutter/flutter/assets/8975114/89b6f674-446d-4f91-80c8-b84e27aa9301)

Pictured Above: Swiping through wondrous application. GPU times are < 2ms and CPU times are under a ms. But frames take 8.33 ms because the drawable isn't available until 1ms before the end of the frame.

Oncreen Texture Usage

Flutter prefers to use the swapchain image as the surface to use for onscreen drawing. A simple Flutter app with limited compositing will often render everything to the swapchain image without any offscreen textures. Compared to offscreen rendering, this reduces memory usage. Which is important because the size of a fullscreen texture on an iPhone 13 Pro is nearly 100MB (2556 * 1179 * 32 bytes), or 200 MB with Wide Gamut, and maybe back to ~100MB if the device supports texture compression. Drawable texture size is NOT attributed to the Flutter application.

## Partial Repaint

"Partial repaint/dirty region management" allows the Flutter engine to reduce the engine/GPU workload when rendering frames that are similar to previous frames. The typical example would be something like a blinking cursor: the engine computes a damage rect and then is able to avoid rendering everything outside of this damage rect, which dramatically reduces CPU/GPU usage. Partial repaint requires the engine to know the id of the drawable resource that will be used before the frame starts in order to compute the damage rect. While there are only three drawables, they are not necessarily cycled in order. This means that partial repaint actually requires that we acquire the drawable as early as possible (from the perspective of the raster thread). 

Partial repaint/Dirty region management was originally filled in https://github.com/flutter/flutter/issues/33939 . I'm not sure if there is a separate design document, but all of the discussion is in that issue.

### Why do we need the drawable ID?

Partial repaint is frequently thought of as a difference between the current frame and the previous frame. But this is wrong, it's the difference between the current frame and the swapchain image we are given. On iOS, this may be any of the previous drawables.

![image](https://github.com/flutter/flutter/assets/8975114/a9d2798d-11d0-4a86-9299-e492dc08f813)

Pictured Above: an application renders a red triangle, then a blue circle, then a green square. Finally, on frame 4 we will render a red triangle again. If the system gives us Drawable A, then we will compute the diff and realize we don't have to do any work! But if we are given B or C, then we still have to re-render. If we don't know the diff before we start rendering, then we can't actually save any work.

### How Important is Partial Repaint?

For the canonical example of the blinking cursor, partial repaint drops CPU usage from 40% down to 15% on iOS. The substantial cost is due to the fade in/out that ticks every single frame instead of once every 100-something ms. See also: https://github.com/flutter/flutter/issues/124526#issuecomment-1512229903 . For other kinds of interaction, like scrolling and page navigation, partial repaint does almost nothing.

It's worth noting that we do not have any other ideas in the pipeline for improving blinking cursor performance without partial repaint enabled. So this change would, in the near term, regress blinking cursor across the board with no amelioration.

# Overview

The structural changes required to defer drawable acquisition are fairly straightforward:

  * impeller::SurfaceMTL should be constructed from a CAMetalLayer and not a Drawable/Texture. The layer already contains the pixel format and size information that is necessary to initialize the render pass descriptors.
  * A special purpose TextureMTL object is created that wraps the CAMetalLayer without acquiring the drawable, exposing a bit that tells the backend it is drawable backed.
  * When the metal texture is requested, the special texture object requests the drawable.
  * RenderPassMTL and BlitPassMTL check for the special drawable blit when configuring the MTLRenderPass/BlitPass descriptors. 
  * Rather than set the descriptor eagerly, the engine waits until it is encoding the cmd buffer (The last possible second).
  * SubmitCommandsAsync checks if the render target texture is the special drawable texture, and only waits on the drawable in the worker thread task.
  * The worker thread task pool is reduced to a single thread to enforce ordering. 

I've built a prototype implementation here: https://github.com/flutter/engine/pull/47976 

![image](https://github.com/flutter/flutter/assets/8975114/f2017656-98b2-4c58-9e52-cb464956a22a)

# Open questions

  * How do we handle error states like failure to acquire a drawable?
  * Can we afford to lose partial repaint?  Is there another path to blinking cursor improvements?
  * Does this improve things for 16.6 Are we vulnerable to another change like 16.6 again?
  * With the removal of partial repaint of Impeller on iOS, the only production backend shipping with it is the legacy iOS Skia renderer. Should we remove it altogether?
    * It was disabled on Skia/Android due to rendering bugs with certain vendors.
    * https://github.com/flutter/flutter/issues/113314 
    * https://github.com/flutter/flutter/issues/105093 
    * Removing partial repaint on Android saved ~1ms per frame on some benchmarks.

  * Do we need to also do this for macOS?
    * macOS drawable acquisition is abstracted away from the embedder, meaning that we can't actually defer it. So we could leave partial repaint running on macOS (though I don't think anyone has wired it up yet)

# Testing plan

Changing how we acquire and present drawables is risky, and unit tests won't be sufficient to cover the change in behavior. Aside from devicelab integration tests and manual tests of some common scenarios, including fast and slow apps, we have a few other tools for reducing risk.

We could land the new drawable acquisition path behind a Plist flag, and then ask certain members of the community to test it. This could be done by opting in during one of the next betas, and then leaving an opt out we can easily cherry pick if there are problems.

I would not ask g3 customers to opt-in or opt-out, instead I would coordinate with the g3 rollers to change the Plist configuration like we've done for Impeller/Wide Gamut already. Then after a release if there are no known issues I would remove the configuration. It's unclear if this should be opt in or opt out, but if it's opt in then we'll need to adjust the devicelab to get coverage.

# Alternatives Considered

  * We can keep partial repaint and remove drawable blocking by unconditionally rendering to an offscreen texture. Then blitting this texture to the onscreen in a worker task. The cost of this is several fold:
    * Small additional overhead of a per-frame blit (hundreds of microseconds)
    * Large additional memory overhead of fullscreen offscreen (100-200MB)
  * No additional memory overhead if there is a backdrop filter present, as it means that we can re-enable "Read from resolve" optimization.
  * Could we do some sort of dynamic switching?

**Comment by jonahwilliams** [comment]
FYI @knopp / @flar  -  I'm not sure if this was discussed during the original partial repaint /DRM design discussion (the thread is a bit hard to follow). But based on my understanding of how partial repaint works on iOS, there is no way to fix drawable aquisition _and_ still get partial repaint. PTAL and correct me if wrong.

**Comment by knopp** [comment]
If we don't know which drawable we're getting upfront, maybe we could merge preexisting dirty areas from all backbuffers. This would result in more area being redrawn as necessary, but for common situations (spinner, cursor, scrollbar) this shouldn't matter much. 

I *think* this should work and it would be better than not doing partial repaint at all, but it's possible that I haven't thought it all the way through and I'm missing something.

**Comment by jonahwilliams** [comment]
Would that require computing three diffs instead of one?

**Comment by knopp** [comment]
No. When computing diff one of the input is preexisting damage for the buffer. If we don't know which buffer we'll be painting to when doing the diff, which can take union of the preexisting damage for both backbuffers.

**Comment by jonahwilliams** [comment]
Ahh, OK - that makes sense. I can experiment with that and see if I can get it to fire for blinking cursor. Though since I typed this out, I did prototype another solution for that which gets even better CPU utilization by letting CA drive the blinking cursor: https://github.com/jonahwilliams/text_platform_view . pure hack.

**Comment by knopp** [comment]
Blinking cursor is one thing - there are other situations - like a spinner, or a smaller scrollable area on iPad.

**Comment by knopp** [comment]
Understood. Though my initial motivation for partial repaint was energy consumption on desktop with an activity indicator. Even though partial repaint is not wired in desktop embedder yet :-/

Pragmatically no implementation of partial repaint is going to beat updating a single cursor CALayer. But I can't say I'm too happy about the idea of repainting entire application for each tiny change.

**Comment by jonahwilliams** [comment]
I'm not certain that we have the drawable problem on macOS, the compositor works very differently there.

**Comment by knopp** [comment]
On macOS we're managing the back buffers [ourselves](https://github.com/flutter/engine/blob/main/shell/platform/darwin/macos/framework/Source/FlutterSurfaceManager.mm). There's no CAMetalLayer. There's no partial repaint because I haven't had time to add it to embedder compositor API and wire it up.

**Comment by jonahwilliams** [comment]
EDIT: i typo'd the bits bytes in my conversion, so the offscreen size is MUCH less: only 12-24 MB!

**Comment by flar** [comment]
On MacOS presumably we are less memory constrained and so maintaining our own offscreen buffer might be more feasible?

**Comment by jonahwilliams** [comment]
We don't actually have partial reapint wired up on macOS. I'm not _actually_ sure if we have a drawable problem there, the compositor/presentation performance characteristics may be completely different. At any rate, the changes here would only directly change iOS for now.

**Comment by flar** [comment]
If they are not cycled in order, does that mean that we can't assume that the total damage is equal to the last N frames? Can it reuse one buffer N+1 times in a row and then the damage recorded for the other buffers is somewhat meaningless at that point as they represent the damage against the (N+1)th frame in the past?

**Comment by knopp** [comment]
Currently presented `frame damage` (current front buffer - previous front buffer) gets added to each of the back buffer accumulated damage. So even if a back buffer has not been used for a while (CAMetalLayer sometimes keeps flipping between just two buffers) the accumulated damage will be correct.

**Comment by flar** [comment]
So then the issue would be if we have a frame that had a big damage and then we go into a cycle of just blinking a cursor, but that frame remains stale then we'd keep using its "big damage" on each partial repaint until the OS decided to throw it back in the mix. Would that happen for many frames in a row, or stochastically we'd get back to it before we did too many frames and then all of the frames would show the same tiny cursor area?

**Comment by knopp** [comment]
> So then the issue would be if we have a frame that had a big damage and then we go into a cycle of just blinking a cursor, but that frame remains stale then we'd keep using its "big damage" on each partial repaint until the OS decided to throw it back in the mix. Would that happen for many frames in a row, or stochastically we'd get back to it before we did too many frames and then all of the frames would show the same tiny cursor area?

That is correct. Back when I was looking at this (couple iOS versions back) most of the time the pattern was regular (all three buffers used), but rarely one buffer was omitted for couple dozen flips. I didn't find any predictable pattern in this, but it always self-corrected quickly so I don't think it would be a problem in practice.

**Comment by lukaszdebowski** [comment]
Is it possible that this is causing a significant scrolling issues with PlatformViews? 
We are using video_player plugin in a ListView, and after upgrading to 3.16 there is a significant load on the raster thread (see screenshot below) even with just 2 platform views (2 video players, not playing)


The WaitForNextDrawable seems to be causing the junk frames ([see image](https://github.com/flutter/flutter/assets/39137870/29157681-0491-46f3-93c4-295ba89e3a15))

**Comment by knopp** [comment]
Sidenote - given how much time we're waiting for drawable, could it mean that we have frame pacing issues that CAMetalLayer is trying to correct?

**Comment by jonahwilliams** [comment]
@lukaszdebowski this is a proposed to change to how Flutter acquires drawables. The existing strategy for acquiring drawables has been consistent for several years. Please file a new bug.

**Comment by jonahwilliams** [comment]
@knopp I think that would only indicate a frame pacing bug if the target time for that frame was before the drawable returned. From what I saw with some experimentation locally, that wasn't the case and the target time was correct. However, the backpressure is delivered in a format that looks like an arbitrary delay which is confusing (and can cause actual issues as I point out above).

**Comment by knopp** [comment]
@jonahwilliams, here's Skia with partial repaint without touching the aquired drawable: https://github.com/knopp/engine/commits/skia_partial_repaint

Couple of remarks: This still aquires the drawable eagerly, preventing that would need undoing some of [this](https://github.com/knopp/engine/commit/4dbdc084442519992673917df3474a3220226c13#diff-94f36bb03816803b4bdc61fa8e259251fd4bc86caa955396f487049eae89e78c). We'd need to use `SkSurfaces::WrapCAMetalLayer` or own version of lazy render target proxy.

The outcome is not completely ideal, `CAMetalLayer` sometimes alternates between two surfaces for several seconds. But I don't think this is a dealbraker and it would be worth looking into whether the pattern changes once surfaces are aquired lazily.

This still doesn't explain why `CAMetalLayer` blocks this long. It might be worth considering not using `CAMetalLayer` at all. We can manage pool of IOSurfaces ourselves and then just set current front buffer as layer contents (similarly to how it is done on macOS). This would also likely improve performance with platform views where overlays are currently extermely heavy (up to 6 buffers (2x CAMetalLayer) per overlay).

**Comment by jonahwilliams** [comment]
> This still doesn't explain why CAMetalLayer blocks this long. It might be worth considering not using CAMetalLayer at all. We can manage pool of IOSurfaces ourselves and then just set current front buffer as layer contents (similarly to how it is done on macOS). This would also likely improve performance with platform views where overlays are currently extermely heavy (up to 6 buffers (2x CAMetalLayer) per overlay).

Because if you look at the presentation time delay in metal system trace, it easily gets into the ~20ms range which implies that we're going to spend some time waiting. Everything on apple documentation I've found (which granted, is not a lot) indicates this is WAI and eagerly acquiring a drawable for any reason is not using the API as intended.

**Comment by jonahwilliams** [comment]
Closing for now as we don't think this will ultimately solve the issue as it has proved to be a dead end.
